File: product-service/internal/services/product_service.go
package services

import (
	"context"
	"fmt"

	"github.com/PedroScheurer/product-service/internal/apperrors"
	"github.com/PedroScheurer/product-service/internal/dtos"
	"github.com/PedroScheurer/product-service/internal/entities"
	"github.com/PedroScheurer/product-service/internal/repositories"
)

type ProductService struct {
	repository                repositories.ProductRepository
	currencyConversionService *CurrencyConversionService
	port                      string
}

func NewProductService(
	repository repositories.ProductRepository,
	currencyConversionService *CurrencyConversionService,
	port string,
) *ProductService {
	return &ProductService{
		repository:                repository,
		currencyConversionService: currencyConversionService,
		port:                      port,
	}
}

func (s *ProductService) FindByID(ctx context.Context, id int64, targetCurrency string) (*dtos.ProductDTO, error) {
	product, err := s.repository.FindByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("find product: %w", err)
	}
	if product == nil {
		return nil, apperrors.NewProductNotFoundError("Produto não encontrado")
	}

	conversionResult := s.currencyConversionService.Convert(product, targetCurrency, s.port)

	return toProductDTO(product, &conversionResult, &targetCurrency), nil
}

func (s *ProductService) FindProductNoConversion(ctx context.Context, id int64) (*dtos.ProductDTO, error) {
	product, err := s.repository.FindByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("find product: %w", err)
	}
	if product == nil {
		return nil, apperrors.NewProductNotFoundError(fmt.Sprintf("Produto não encontrado com o ID: %d", id))
	}

	environment := "Product-service running on port: " + s.port

	return toProductDTO(product, &dtos.ConversionResult{Environment: environment}, nil), nil
}

func (s *ProductService) FindProductsPaged(ctx context.Context, targetCurrency string, page, size int, sortBy, sortDir string) (*dtos.Page[dtos.ProductDTO], error) {
	products, total, err := s.repository.FindAllPaged(ctx, page, size, sortBy, sortDir)
	if err != nil {
		return nil, fmt.Errorf("find products paged: %w", err)
	}

	content := make([]dtos.ProductDTO, 0, len(products))
	for i := range products {
		product := products[i]
		conversionResult := s.currencyConversionService.Convert(&product, targetCurrency, s.port)
		content = append(content, *toProductDTO(&product, &conversionResult, &targetCurrency))
	}

	return dtos.NewPage(content, page, size, total), nil
}

func toProductDTO(product *entities.ProductEntity, conversionResult *dtos.ConversionResult, requestedCurrency *string) *dtos.ProductDTO {
	var convertedPrice *float64
	var environment *string

	if conversionResult != nil {
		environment = &conversionResult.Environment
		if requestedCurrency != nil {
			price := conversionResult.ConvertedPrice
			convertedPrice = &price
		}
	}

	return &dtos.ProductDTO{
		ID:                product.ID,
		Name:              product.Name,
		Instructor:        product.Instructor,
		ImageURL:          product.ImageURL,
		VideoURL:          product.VideoURL,
		Description:       product.Description,
		Workload:          product.Workload,
		Modules:           product.Modules,
		Price:             product.Price,
		Currency:          product.Currency,
		ConvertedPrice:    convertedPrice,
		RequestedCurrency: requestedCurrency,
		Environment:       environment,
	}
}

File: product-service/src/main/java/br/edu/atitus/productservice/services/ProductService.java
package br.edu.atitus.productservice.services;

import br.edu.atitus.productservice.dtos.ConversionResult;
import br.edu.atitus.productservice.dtos.ProductDTO;
import br.edu.atitus.productservice.entities.ProductEntity;
import br.edu.atitus.productservice.infrastructure.exceptions.ProductNotFoundException;
import br.edu.atitus.productservice.repositories.ProductRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
public class ProductService {

    private final ProductRepository repository;
    private final CurrencyConversionService currencyConversionService;

    @Value("${server.port}")
    private String port;

    public ProductService(ProductRepository repository, CurrencyConversionService currencyConversionService) {
        this.repository = repository;
        this.currencyConversionService = currencyConversionService;
    }

    public ProductDTO findById(Long id, String targetCurrency) {
        ProductEntity product = repository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException("Produto não encontrado"));

        ConversionResult conversionResult = currencyConversionService.convert(product, targetCurrency, port);

        return new ProductDTO(product.getId(),
                product.getDescription(),
                product.getBrand(),
                product.getModel(),
                product.getPrice(),
                product.getCurrency(),
                product.getStock(),
                conversionResult.environment(),
                conversionResult.convertedPrice(),
                targetCurrency);
    }

    public ProductDTO findProductNoConversion(Long idProduct) {
        ProductEntity product = repository.findById(idProduct)
                .orElseThrow(() -> new ProductNotFoundException("Produto não encontrado com o ID: " + idProduct));

        return new ProductDTO(
                product.getId(),
                product.getDescription(),
                product.getBrand(),
                product.getModel(),
                product.getPrice(),
                product.getCurrency(),
                product.getStock(),
                "Product-service running on port: " + port,
                null,
                null
        );
    }

    public Page<ProductDTO> findProductsPaged(String targetCurrency, Pageable pageable) {
        Page<ProductEntity> products = repository.findAll(pageable);

        return products.map(product -> {
            ConversionResult conversionResult = currencyConversionService.convert(product, targetCurrency, port);

            return new ProductDTO(product.getId(),
                    product.getDescription(),
                    product.getBrand(),
                    product.getModel(),
                    product.getPrice(),
                    product.getCurrency(),
                    product.getStock(),
                    conversionResult.environment(),
                    conversionResult.convertedPrice(),
                    targetCurrency);
        });
    }
}
